client/fullnode_cache.go

```diff
@@ -0,0 +1,219 @@
+	//   - 10 is a good balance for most workloads
+	numShards = 10
+
+	// evictionPercentage:
+	//   - % of LRU entries evicted per shard when full
+	//   - 10% = incremental cleanup, avoids memory spikes
+	//   - SturdyC also evicts expired entries in background
+	evictionPercentage = 10
+
+	// TODO_TECHDEBT(@commoddity): See Issue #291 for improvements to refresh logic
+	// minEarlyRefreshPercentage:
+	//   - Earliest point (as % of TTL) to start background refresh
+	//   - 0.75 = 75% of TTL (e.g. 22.5s for 30s TTL)
+	minEarlyRefreshPercentage = 0.75
+
+	// maxEarlyRefreshPercentage:
+	//   - Latest point (as % of TTL) to start background refresh
+	//   - 0.9 = 90% of TTL (e.g. 27s for 30s TTL)
+	//   - Ensures refresh always completes before expiry
+	maxEarlyRefreshPercentage = 0.9
+)
+
+// pubKeyCacheTTL: No TTL for the account public key cache since account data never changes.
+//
+// time.Duration(math.MaxInt64) equals ~292 years, which is effectively infinite.
+const pubKeyCacheTTL = time.Duration(math.MaxInt64)
+
+// getCacheDelays returns the min/max delays for SturdyC's Early Refresh strategy.
+//   - Proactively refreshes cache before expiry (prevents misses/latency spikes)
+//   - Refresh window: 75-90% of TTL (e.g. 22.5-27s for 30s TTL)
+//   - Spreads requests to avoid thundering herd
+//
+// See: https://github.com/viccon/sturdyc?tab=readme-ov-file#early-refreshes
+func getCacheDelays(ttl time.Duration) (min, max time.Duration) {
+	minFloat := float64(ttl) * minEarlyRefreshPercentage
+	maxFloat := float64(ttl) * maxEarlyRefreshPercentage
+
+	// Round to the nearest second
+	min = time.Duration(minFloat/float64(time.Second)+0.5) * time.Second
+	max = time.Duration(maxFloat/float64(time.Second)+0.5) * time.Second
+	return
+}
+
+// Prefix for cache keys to avoid collisions with other keys.
+const (
+	sessionCacheKeyPrefix       = "session"
+	accountPubKeyCacheKeyPrefix = "pubkey"
+)
+
+// cachingFullNode wraps a LazyFullNode with SturdyC-based caching.
+//   - Early refresh: background updates before expiry (prevents thundering herd/latency spikes)
+//   - Example: 30s TTL, refresh at 22.5–27s (75–90%)
+//   - Benefits: zero-latency reads, graceful degradation, auto load balancing
+//
+// Docs: https://github.com/viccon/sturdyc
+type fullNodeWithCache struct {
+	logger polylog.Logger
+
+	// Underlying node for protocol data fetches
+	underlyingFullNode *fullNode
+
+	// Session cache
+	// TODO_MAINNET_MIGRATION(@Olshansk): Revisit after mainnet
+	sessionCache *sturdyc.Client[sessiontypes.Session]
+
+	// The account public key cache; used to cache account public keys indefinitely.
+	// It has an infinite TTL and is populated only once on startup.
+	accountPubKeyCache *sturdyc.Client[cryptotypes.PubKey]
+}
+
+// newFullNodeWithCache wraps a fullNode with:
+//   - Session cache: refreshes early to avoid thundering herd/latency spikes
+//   - Account public key cache: indefinite cache for account data
+func newFullNodeWithCache(
```

Comment:
I don't think fullnode* structs should be private: making them public and allowing the user (e.g. PATH) to customize any required aspect e.g. GRPC settings, makes more sense.

Also, please consider the approach of making fullnode* structs useful on their own: e.g. a reliable way of interacting with the protocol for purposes other than running a gateway.